<?xml version="1.0" encoding="UTF-8"?>
<!-- Launch autonomous control for the EZ-RASSOR.
     Written by Cameron Taylor and Tiger Sachse. -->
<launch>
  <arg name="wheel_instructions_topic"/>
  <arg name="front_arm_instructions_topic"/>
  <arg name="back_arm_instructions_topic"/>
  <arg name="front_drum_instructions_topic"/>
  <arg name="back_drum_instructions_topic"/>
  <arg name="digsite_x_coord" default="10"/>
  <arg name="digsite_y_coord" default="10"/>
  <arg name="spawn_x_coord" default="0"/>
  <arg name="spawn_y_coord" default="0"/>
  <arg name="max_linear_velocity" default="0.125"/>
  <arg name="max_angular_velocity" default="0.3"/>
  <arg name="enable_real_odometry" default="false"/>
  <arg name="swarm_control" default="false"/>
  <arg name="obstacle_threshold" default="4.0"/>
  <arg name="obstacle_buffer" default="1.5"/>
  <arg name="move_increment" default="3.0"/>
  <arg name="max_obstacle_angle" default="45.0"/>
  <arg name="min_hole_diameter" default="3.0"/>

  <!-- Launch the autonomous control node. -->
  <node pkg="ezrassor_autonomous_control"
        name="autonomous_control"
        type="autonomous_control"
        output="screen">
    <param name="digsite_x_coord" value="$(arg digsite_x_coord)"/>
    <param name="digsite_y_coord" value="$(arg digsite_y_coord)"/>
    <param name="spawn_x_coord" value="$(arg spawn_x_coord)"/>
    <param name="spawn_y_coord" value="$(arg spawn_y_coord)"/>
    <param name="max_linear_velocity" value="$(arg max_linear_velocity)"/>
    <param name="max_angular_velocity" value="$(arg max_angular_velocity)"/>
    <param name="enable_real_odometry" value="$(arg enable_real_odometry)"/>
    <param name="wheel_instructions_topic"
           value="$(arg wheel_instructions_topic)"/>
    <param name="front_arm_instructions_topic"
           value="$(arg front_arm_instructions_topic)"/>
    <param name="back_arm_instructions_topic"
           value="$(arg back_arm_instructions_topic)"/>
    <param name="front_drum_instructions_topic"
           value="$(arg front_drum_instructions_topic)"/>
    <param name="back_drum_instructions_topic"
           value="$(arg back_drum_instructions_topic)"/>
    <param name="swarm_control"
           value="$(arg swarm_control)"/>
    <param name="obstacle_threshold"
           value="$(arg obstacle_threshold)"/>
    <param name="obstacle_buffer"
           value="$(arg obstacle_buffer)"/>
    <param name="move_increment"
           value="$(arg move_increment)"/>
  </node>

  <node pkg="image_view" type="image_view" name="image_view">
	<remap from="image" to="depth/image_raw"/>
  </node>

  <node pkg="ezrassor_autonomous_control" type="obstacle_detection" name="obstacle_detection">
       <param name="max_angle"
              value="$(arg max_obstacle_angle)"/>
       <param name="max_obstacle_dist"
              value="$(arg obstacle_threshold)"/>
       <param name="min_hole_diameter"
              value="$(arg min_hole_diameter)"/>
  </node>

  <group if="$(arg enable_real_odometry)">
       <node pkg="rtabmap_ros" type="rgbd_odometry" name="visual_odometry">
              <param name="frame_id" value="depth_camera_optical_frame"/>
              <param name="publish_tf" value="false"/>
              <param name="publish_null_when_lost" value="false"/>
              <param name="Odom/ResetCountdown" value="1"/>
              <remap from="rgb/image" to="color/image_raw"/>
              <remap from="rgb/camera_info" to="color/camera_info"/>
              <remap from="depth/image" to="depth/image_raw"/>
              <remap from="odom" to="visual_odom/odom"/>
              <remap from="odom_info" to="visual_odom/odom_info"/>
              <remap from="odom_last_frame" to="visual_odom/odom_last_frame"/>
              <remap from="odom_local_map" to="visual_odom/odom_local_map"/>
       </node>

       <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
              <rosparam command="load" file="$(find ezrassor_autonomous_control)/config/ekf_template.yaml" />
       </node>
  </group>
</launch>
